workspaceitems.md

```diff
@@ -0,0 +1,113 @@
+# WorkspaceItem Endpoints
+[Back to the list of all defined endpoints](endpoints.md)
+
+## Main Endpoint
+**/api/submission/workspaceitems**   
+
+Provide access to the workspaceitems. It returns the list of existent workspaceitems.
+
+Example: to be provided
+
+## Single Item
+**/api/submission/workspaceitems/<:id>**
+
+Provide detailed information about a specific workspaceitem. The JSON response document is as follow
+```json
+{
+  "id": 1911,
+  "lastModified": "2017-06-24T00:40:54.970+0000",
+  "sections": {
+  	 "collection": "05457c63-b392-4629-a373-f2d66ee9ee33",
+  	 "traditional-page1": {
+  	 	"dc.title" : [{value: "Sample Submission Item"}],
+  	 	"dc.contributor.author" : [
+  	 		{value: "Bollini, Andrea", authority: "rp00001", confidence: 600}
+  	 	]
+  	 },
+  	 "traditional-page2": {
+  	 	"dc.subject" : [{value: "Test"}, {value: "JSON"}, {value: "REST"}],
+  	 	"dc.description.abstract" : [
+  	 		{value: "This is a long multiline text\n This is a second line of the abstract"}
+  	 	]
+  	 },
+  	 "license": {
+  	 	acceptanceDate: "2017-06-24T00:40:54.970+0000",
+  	 	url: "http://dspace7.4science.it/api/core/bitstreams/8d33bdfb-e7ba-43e6-a93a-f445b7e8a1e2/content"
+  	 },
+  	 "uploads": [ 
+  	 	{
+  	 		metadata: {
+  	 			"dc.title" : [{value: "sample_file.pdf"}],
+  	 			"dc.description" : [{value: "Description of the sample file"}]
```

Comment:
Just to clarify the differences between the two formats:

# Old format
```
"metadata": [
    {
      "key": "dc.contributor.author",
      "value": "Bollini, Andrea",
      "language": "en",
      "authority": "rp00001",
      "confidence": 600
    },
    {
      "key": "dc.contributor.author",
      "value": "Desair, Tom"
    }
]
```
The JSON patch commands would be:
Add a new author:
```
{
  "op": "add",
  "path": "/metadata/-",
  "value": {
    "key": "dc.contributor.author",
    "value": "John, Doe",
    "language": "en"
  }
}
```

Remove third metadata value
```
{
  "op": "remove",
  "path": "/metadata/2"
}
```

Update the value of the second metadata field
```
{
  "op": "replace",
  "path": "/metadata/1/value",
  "value": "Doe, John"
}
```

# New format
```
"metadata" : {
  "dc.contributor.author" : [
    {
      "value": "Bollini, Andrea",
      "language": "en",
      "authority": "rp00001",
      "confidence": 600
    },
    {
      "value": "Desair, Tom",
      "language": "en"
    }    
  ]
}
```
The JSON Patch commands would be:
Add a new author:
```
{
  "op": "add",
  "path": "/metadata/dc.contributor.author/-",
  "value": {
    "value": "Doe, John",
    "language": "en"
  }
}
```

Remove the third author metadata value
```
{
  "op": "remove",
  "path": "/metadata/dc.contributor.author/2"
}
```

Update the value of the second author
```
{
  "op": "replace",
  "path": "/metadata/dc.contributor.author/1/value",
  "value": "Doe, John"
}
```

To get all metadata values of object:
* In the old format, one has to loop over all objects in the `metadata` array
* In the new format, one has to loop over all the fields in the `metadata` object.

So while not strictly necessary to implement JSON patch or JSON Merge Patch, the new format does group metadata values together and prevents repetition of the metadata field key. 

So for me this is OK. 👍

Reply:
Keep in mind that it could be also impossible for the client to get the full list of metadata of an item for security constraints (some metadata can be hidden in the submission). Of course the server side can take in account that but it is much complex than just allow easy navigation of the metadata by "field name"